File: server/Models/Habit.cs
using System.ComponentModel.DataAnnotations;

namespace TasksTrack.Models
{
    public class Habit
    {
        public int Id { get; set; }

        [Required(ErrorMessage = "Name is required")]
        [StringLength(200, ErrorMessage = "Name cannot exceed 200 characters")]
        public required string Name { get; set; }

        [StringLength(1000, ErrorMessage = "Description cannot exceed 1000 characters")]
        public string? Description { get; set; }

        [Required(ErrorMessage = "Metric type is required")]
        [StringLength(50, ErrorMessage = "Metric type cannot exceed 50 characters")]
        public required string MetricType { get; set; } // "minutes", "miles", "reps", "pages", etc.

        [StringLength(20, ErrorMessage = "Unit cannot exceed 20 characters")]
        public string? Unit { get; set; } // Display unit like "min", "mi", "reps", "pages"

        public decimal? Target { get; set; } // Daily/weekly target value
        public string? TargetFrequency { get; set; } // "daily", "weekly"

        [StringLength(100, ErrorMessage = "Category cannot exceed 100 characters")]
        public string? Category { get; set; } // "Health", "Learning", "Creative", etc.

        public bool IsActive { get; set; } = true;
        public DateTimeOffset CreatedDate { get; set; }
        public DateTimeOffset? UpdatedDate { get; set; }

        [Required(ErrorMessage = "CreatedBy is required")]
        public required string CreatedBy { get; set; }

        public string? UpdatedBy { get; set; }

        // Color for UI display
        [StringLength(7, ErrorMessage = "Color cannot exceed 7 characters")]
        public string? Color { get; set; } // Hex color code

        // Icon for UI display
        [StringLength(50, ErrorMessage = "Icon cannot exceed 50 characters")]
        public string? Icon { get; set; }
    }
}

File: client/src/types/habit.ts
import { z } from 'zod';
import { getUnitsForMetricType } from './constants';
import { type MetricType } from '@/utils/unitStandardization';

// Habit entity schema (complete database model) - matches Habit.cs exactly
export const habitEntitySchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string().nullable(),
  metricType: z.string(),
  unit: z.string().nullable(),
  target: z.number().nullable(), // decimal? in C# becomes number | null
  targetFrequency: z.string().nullable(),
  category: z.string().nullable(),
  isActive: z.boolean(),
  createdDate: z.string(), // DateTimeOffset serialized as string
  updatedDate: z.string().nullable(), // DateTimeOffset? serialized as string or null
  createdBy: z.string(),
  updatedBy: z.string().nullable(),
  color: z.string().nullable(), // Hex color code
  icon: z.string().nullable(), // Icon name for UI
});

// Habit form schema (for creating/editing habits)
export const habitFormSchema = habitEntitySchema.pick({
  name: true,
  description: true,
  metricType: true,
  unit: true,
  target: true,
  targetFrequency: true,
  category: true,
  color: true,
  icon: true,
}).superRefine((data, ctx) => {
  // Custom validation: if unit is provided, validate it against the metric type
  if (data.unit && data.unit.length > 0) {
    try {
      const validUnits = getUnitsForMetricType(data.metricType as MetricType);
      if (validUnits.length > 0 && !validUnits.includes(data.unit)) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: `Unit "${data.unit}" is not valid for metric type "${data.metricType}". Valid units are: ${validUnits.join(', ')}`,
          path: ['unit']
        });
      }
    } catch (error) {
      // If metricType is not a valid MetricType, skip unit validation
      // This allows for flexibility with backend data that might have different metric types
    }
  }

  // Custom validation: if target is provided and > 0, targetFrequency should be provided
  if (data.target && data.target > 0 && !data.targetFrequency) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: 'Target frequency is required when target is set',
      path: ['targetFrequency']
    });
  }
});

// Export types
export type IHabit = z.infer<typeof habitEntitySchema>;
export type HabitFormData = z.infer<typeof habitFormSchema>;

// Request/Response types for API operations
export interface IHabitCreateRequest {
  name: string;
  description?: string;
  metricType: string;
  unit?: string;
  target?: number;
  targetFrequency?: string;
  category?: string;
  color?: string;
  icon?: string;
}

export interface IHabitUpdateRequest {
  name?: string;
  description?: string;
  metricType?: string;
  unit?: string;
  target?: number;
  targetFrequency?: string;
  category?: string;
  color?: string;
  icon?: string;
}
